### PR TITLE
Simplify global range constraints

### LINKED ISSUE #449: Witness generation: Global constraints should not use `PolynomialReference`
We currently store global range constraints in a `BTreeMap<&PolynomialReference, RangeConstraint<T>>`. However, a polynomial reference contains the `next` field, so there could be a global constraint for both `X` and `X'`, which does not make sense.

In practice, next reference typically don't appear in identities that enforce global range constraints. But this also means that when we have a range constraint for `X`, but look it up for `X'`, we get `None`.

### PR DESCRIPTION
**Pull request by georgwiese**
Fixes #449

Global range constraints are now indexed by the *witness* polynomial ID (`PolyID`) instead of `&PolynomialReference`. This has the following consequences:
- There are no more range constraints for constant columns (should not be needed)
- The `next` flag is ignored (which was actually a bug, as described in the issue)
- The types are a lot simpler now (at the expense of not easily being able to print the polynomial's name anymore)

### CODE REVIEW

File: executor/src/witgen/generator.rs
Comment (leonardoalt) [inline executor/src/witgen/generator.rs:26]: Maybe `BTreeMap<u64, RangeConstraint<T>>` could be extracted as a type so it doesn't need to be repeated everywhere? Not too important, just a readability suggestion.

**Review by leonardoalt — COMMENTED** [review]
lgtm, maybe @Schaeff should take a look too

File: executor/src/witgen/generator.rs
Comment (georgwiese) [reply executor/src/witgen/generator.rs:26]: Hmm, I actually find it more readable this way, because I don't have to remember a new type and that it's just a map. That might be a matter of taste though; I've realized that Rust folks like to introduce new types a lot 😄

File: executor/src/witgen/global_constraints.rs
Comment (leonardoalt) [reply executor/src/witgen/global_constraints.rs:86]: I think that was also @georgwiese 's idea in the first place, wasn't it?

Comment (Schaeff) [reply executor/src/witgen/global_constraints.rs:86]: (not a blocker for this PR)

File: executor/src/witgen/generator.rs
Comment (georgwiese) [reply executor/src/witgen/generator.rs]: With that I actually agree. There are several places in witgen where `usize` is the key, and it's not clear whether it's supposed to be an ID or some sort of index.

After a few iterations, I figured it's best to use the existing `PolyID`. It also contains the type (committed or fixed), which is actually what we want, because the `u64` ID numbers will collide between the types!

I renamed `PolynomialReference::poly_id()` -> `PolynomialReference::poly_id_u64()` and added `PolynomialReference::poly_id()` for convenience. I'll remove `poly_id_u64()` in #412, to keep the changes of this PR small.

File: executor/src/witgen/global_constraints.rs
Comment (georgwiese) [reply executor/src/witgen/global_constraints.rs:86]: Hmm, few thoughts on this:
- I didn't dive into the code that computes global range constraints here, but it's kinda weird that the other functions work on Polynomial references instead of `PolyID`. For example, I guess defining a bit constraint as `X' * (1 - X') = 0` would break the code. What you'd *really* need to do is to detect range constraints per reference, group by ID and intersect the ranges.
- What's nice about a `PolynomialReference` is that it can pretty-print itself because it has (i.e. owns) the name of the polynomial. Because of that, we use `&PolynomialReference` as keys in various places...

I think what would be nice is to have something like:
```
struct Polynomial:
  name: &'static str,
  id: PolyID
```

or

```
struct Polynomial<'a>:
  name: &'a str,
  id: PolyID
```

The first approach involves leaking the column names, which I think would be fine and prevent using convoluting the code with lifetime annotations...

What do you think?

**Review by Schaeff — APPROVED** [review]
Lgtm! Please squash before merge :)

**Pull request merged**
